### PR TITLE
Batch processing

### PR DESCRIPTION
**Pull request by giang-nghg**
This PR adds the batch processing capability to ivcfmerge, so that it can process large number of input files that exceed the OS' open file limit.

Updated README:

# ivcfmerge: Incremental VCF merge

## Purpose

We provides a utility to merge a large number of VCF files (possibly too many to open at once) incrementally, that only use as much memory as each line in one of the input files takes (so the fewer samples you have in each of the input files, the less memory you need).

#### If the number of input files is small (can be opened all at once)

```python
from contextlib import ExitStack
from ivcfmerge import ivcfmerge

filenames = [...]    # List/iterator of relative/absolute paths to input files
output_path = '...'  # Where to write the merged VCF to

with ExitStack() as stack:
    files = map(lambda fname: stack.enter_context(open(fname)), filenames)
    with open(output_path) as outfile:
        ivcfmerge(files, outfile)
```

#### If the number of input files is big (cannot be opened all at once)

```python
from ivcfmerge.batch import ivcfmerge_batch

filenames = [...]    # List/iterator of relative/absolute paths to input files
output_path = '...'  # Where to write the merged VCF to
batch_size = 1000    # How many files to open and merge at once

ivcfmerge_batch(filenames, output_path, batch_size)
```

## Batch size

An important parameter for this utility is `batch_size`, which indicates how many files to open and merge each batch,
since the total number of input files can exceed the number of open files allowed by the OS.

The default value for this parameter is 1000.

### CODE REVIEW

File: tests/test_ivcfmerge_batch.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+from tempfile import NamedTemporaryFile
+
+from hypothesis import given, strategies as st
+
+from ivcfmerge.batch import ivcfmerge_batch
+
+
+@given(batch_size=st.integers(min_value=2))
+def test_merging_example_input(batch_size):
+    input_paths = [
```
Comment (Zhicheng-Liu) [inline tests/test_ivcfmerge_batch.py]: Can we also add a test case for handling empty input paths list?

File: ivcfmerge/batch.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import shutil
+from contextlib import ExitStack
+from queue import Queue
+from tempfile import NamedTemporaryFile
+
+from . import ivcfmerge
+
+
+def ivcfmerge_batch(input_paths, output_path, batch_size=1000):
```
Comment (Zhicheng-Liu) [inline ivcfmerge/batch.py]: Please expose this method also.

File: tests/test_ivcfmerge_batch.py (hunk 1 above)
Comment (Zhicheng-Liu) [reply tests/test_ivcfmerge_batch.py]: Please also add a test case for handling list that contains only one input path.

Comment (giang-nghg) [reply tests/test_ivcfmerge_batch.py]: What do you want the behavior to be when the input list is empty?

File: ivcfmerge/batch.py (hunk 2 above)
Comment (giang-nghg) [reply ivcfmerge/batch.py]: I already did

File: tests/test_ivcfmerge_batch.py (hunk 1 above)
Comment (giang-nghg) [reply tests/test_ivcfmerge_batch.py]: by default, `ivcfmerge` (no batch) write an empty file with an empty input list, should we maintain that behavior for batch version?

Comment (Zhicheng-Liu) [reply tests/test_ivcfmerge_batch.py]: I think when the input list is empty, the function will produce an empty output file. This is probably what the current implementation already does. When the input list contains only one file, the function should produce an output file exactly the same as the input file.

File: ivcfmerge/batch.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    if batch_size < 2 or batch_size > len(input_paths):
+        batch_size = len(input_paths)
+
+    if batch_size == len(input_paths):
+        batch = input_paths
+    else:
+        work_queue = Queue()
+        [work_queue.put(item) for item in input_paths]
```
Comment (Zhicheng-Liu) [inline ivcfmerge/batch.py]: I see no benefit to use list comprehension here over simple for loop. In fact it will create a temporary list which contains `len(input_paths)` `None` elements.

File: ivcfmerge/batch.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            batch = []
+            while not work_queue.empty() and len(batch) < batch_size:
+                batch.append(work_queue.get_nowait())
+
+            if len(batch) < batch_size or work_queue.empty():
+                break
+
+            tmp_filename = NamedTemporaryFile().name
+            _ivcfmerge(batch, tmp_filename)
+            work_queue.put(tmp_filename)
```
Comment (Zhicheng-Liu) [inline ivcfmerge/batch.py]: I have another thought on the ordering of samples in the merged VCFs. I think it is reasonable for a user to expect that the order will remain the same as they appear in the input. This requirement is not listed in the ticket. I will add another ticket to meet this requirement.
